src/Lib.P2P/Transports/Tcp.cs

```diff
@@ -0,0 +1,293 @@
+                socket?.Dispose();
+                cancel.ThrowIfCancellationRequested();
+            }
+
+            var timeout = (int) Math.Max(MinReadTimeout.TotalMilliseconds, latency.TotalMilliseconds * 3);
+            socket.LingerState = new LingerOption(false, 0);
+            socket.ReceiveTimeout = timeout;
+            socket.SendTimeout = timeout;
+            Stream stream = new NetworkStream(socket, true);
+            stream.ReadTimeout = timeout;
+            stream.WriteTimeout = timeout;
+
+            stream = new DuplexBufferedStream(stream);
+
+            if (!cancel.IsCancellationRequested)
+            {
+                return stream;
+            }
+            
+            Log.Trace("cancel " + address);
+            await stream.DisposeAsync();
+            cancel.ThrowIfCancellationRequested();
+
+            return stream;
+        }
+
+        /// <inheritdoc />
+        public MultiAddress Listen(MultiAddress address,
+            Action<Stream, MultiAddress, MultiAddress> handler,
+            CancellationToken cancel)
+        {
+            var port = address.Protocols
+               .Where(p => p.Name == "tcp")
+               .Select(p => int.Parse(p.Value))
+               .FirstOrDefault();
+            
+            var ip = address.Protocols
+               .FirstOrDefault(p => p.Name == "ip4" || p.Name == "ip6");
+            
+            if (ip == null)
+            {
+                throw new ArgumentException($"Missing IP address in '{address}'.", nameof(address));
+            }
+            
+            var ipAddress = IPAddress.Parse(ip.Value);
+            var endPoint = new IPEndPoint(ipAddress, port);
+            var socket = new Socket(
+                endPoint.AddressFamily,
+                SocketType.Stream,
+                ProtocolType.Tcp);
+            try
+            {
+                socket.Bind(endPoint);
+                socket.Listen(100);
+            }
+            catch (Exception e)
+            {
+                socket.Dispose();
+                throw new Exception("Bind/listen failed on " + address, e);
+            }
+
+            // If no port specified, then add it.
+            var actualPort = ((IPEndPoint) socket.LocalEndPoint).Port;
+            
+            if (port != actualPort)
+            {
+                address = address.Clone();
+                var protocol = address.Protocols.FirstOrDefault(p => p.Name == "tcp");
+                if (protocol != null)
+                {
+                    protocol.Value = actualPort.ToString();
+                }
+                else
+                {
+                    address.Protocols.AddRange(new MultiAddress("/tcp/" + actualPort).Protocols);
+                }
+            }
+
+            _ = Task.Run(() => ProcessConnection(socket, address, handler, cancel), cancel);
+
+            return address;
+        }
+
+        private static void ProcessConnection(Socket socket,
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Refactor this method to reduce its Cognitive Complexity from 16 to the 15 allowed.](https://app.codacy.com/manual/catalyst-network/Catalyst.Framework/pullRequest?prid=4832783)